### ISSUE TITLE
[OSS] Expose `clip_grad_norm` for the sharded state

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/fairscale#98]
## 🚀 Feature
Expose a method capable of returning the grad norm and clipping them which works on the sharded state

## Motivation
Similar to https://pytorch.org/docs/master/generated/torch.nn.utils.clip_grad_norm_.html but with OSS and sharded gradients

## Pitch
Could be done within ShardedDataParallel, reduce the norms and normalize the shards

## Alternatives
Not exposing this, could be problematic for some usecases

## Additional context
Exposed in a similar context by fairseq_optimizer

**Comment by blefaudeux** [comment]
@stephenroller having a look at that right now, would you think that it makes more sense to expose it in the sharded optimizer or in the DDP engine ? the torch utils implementation is purely functional, but that would probably not be doable here

**Comment by stephenroller** [comment]
It's commonly exposed in the fp16 context (apex). I'd say the optimizer might make more sense in this case? Assuming most people do clipping near the same place they do the optimizer.step. I'm sure I could deal with both reasonably as a user though

**Comment by blefaudeux** [comment]
Sounds good, thanks for the quick feedback !

**Comment by blefaudeux** [comment]
adding notes to document here, apex (fp16) actually exposes the optimizer's parameters so that the normal torch util grad clipping can be used, see https://nvidia.github.io/apex/advanced.html#gradient-clipping.
It feels a bit different from the DDP usecase though, because in that case each of the shards will be normalized independently, unless the user does that in several steps (collect max norm for each shard, sync the max across ranks, normalize each shard)

**Comment by blefaudeux** [comment]
> adding notes to document here, apex (fp16) actually exposes the optimizer's parameters so that the normal torch util grad clipping can be used, see https://nvidia.github.io/apex/advanced.html#gradient-clipping.
> It feels a bit different from the DDP usecase though, because in that case each of the shards will be normalized independently, unless the user does that in several steps (collect max norm for each shard, sync the max across ranks, normalize each shard)

aah of course, I just realized that in Apex's case the optimizer is not sharded, my bad. So handling the inter-rank sync is specific to us, needs a "new" solution
